### PR TITLE
Improve TypeScript safety

### PR DESCRIPTION
**Pull request by TimWolla**
**NOT TESTED YET**

### CODE REVIEW

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Date/Picker.ts [hunk 1]
```diff
@@ -698,7 +698,7 @@ const DatePicker = {
             value = "";
           } else {
             element.dataset.value = time.toString();
-            const format = isTimeOnly ? "formatTime" : "formatDate" + (isDateTime ? "Time" : "");
+            const format = isTimeOnly ? "formatTime" : isDateTime ? "formatDateTime" : "formatDate";
```
Comment (dtdesign) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/Date/Picker.ts:701]: Can't we just split this up into the separate calls? The performance is the same, but the viewer no longer barfs internally.

File: wcfsetup/install/files/ts/WoltLabSuite/Core/I18n/Plural.ts [hunk 2]
```diff
@@ -9,184 +9,105 @@
...
-  },
-
-  /**
-   * `v` represents the number of digits of the fractional part (1.234 yields 3)
-   */
-  getV(n: number): number {
-    return n.toString().replace(/^[^.]*\.?/, "").length;
-  },
-
+enum Category {
```
Comment (dtdesign) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/I18n/Plural.ts:12]:
```suggestion
const enum Category {
```
This allows the compiler to inline the value and remove the enum from the final code. There is no partical use of carrying it around on runtime?

**Review by dtdesign — COMMENTED** [review]
In general I would also like to see enforced return types on any function that isn't anonymous. Typical callbacks are fine because the return value of the function is inferred from the context anyway.

File: wcfsetup/install/files/ts/WoltLabSuite/Core/I18n/Plural.ts (hunk 2 above)
Comment (TimWolla) [reply wcfsetup/install/files/ts/WoltLabSuite/Core/I18n/Plural.ts:12]: I attempted that, but somehow that caused issues with defining the `Parameters` type. I've re-tested this and apparently my later changes fixed that definition issue.

**Comment by TimWolla** [comment]
> In general I would also like to see enforced return types on any function that isn't anonymous.

eslint already does this at module boundaries (i.e. exported functions). I would need to check whether there's anything to require this for all named functions.

**Comment by TimWolla** [comment]
I now verified the correct workings of the `Plural` module which received a bigger refactoring. The other modules just got some types, so this PR is ready now.

**Review by Gravatronics — APPROVED** [review]

**Pull request merged**